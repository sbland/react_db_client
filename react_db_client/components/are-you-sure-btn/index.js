export { AreYouSureBtn } from './are-you-sure-btn';
