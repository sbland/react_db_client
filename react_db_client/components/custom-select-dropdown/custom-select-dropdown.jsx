import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';

import './customSelectDropdown.scss';
import { DropDownItem } from './drop-down-item';

export const CustomSelectDropdown = ({
  options,
  handleSelect,
  isOpen,
  handleClose,
  firstItemRef,
  goBackToSearchField = () => {},
}) => {
  return <div>Hello</div>
  const menuRef = useRef(null);
  const itemRefs = useRef([]);
  const [currentItemFocus, setCurrentItemFocus] = useState(0);

  useEffect(() => {
    itemRefs.current = itemRefs.current.slice(0, options.length);
  }, [options]);

  function handleClickOutside(event) {
    if (isOpen && menuRef.current && !menuRef.current.contains(event.target)) {
      handleClose();
    }
  }

  // TODO: Disabled as catches user input
  // useEffect(() => {
  //   if (isOpen) {
  //     // focus on first item when dropdown opened
  //     if (itemRefs.current[0]) itemRefs.current[0].focus();
  //   }
  // }, [isOpen]);

  useEffect(() => {
    // Bind the event listener
    document.addEventListener('mousedown', handleClickOutside);
    return () => {
      // Unbind the event listener on clean up
      document.removeEventListener('mousedown', handleClickOutside);
    };
  });

  const handleItemKeyDown = (e) => {
    if (e.key === 'ArrowDown' && currentItemFocus < options.length - 1) {
      e.preventDefault();
      itemRefs.current[currentItemFocus + 1].focus();
      setCurrentItemFocus((prev) => prev + 1);
    }
    if (e.key === 'ArrowUp' && currentItemFocus > 0) {
      e.preventDefault();
      itemRefs.current[currentItemFocus - 1].focus();
      setCurrentItemFocus((prev) => prev - 1);
    }
    if (e.key === 'ArrowUp' && currentItemFocus === 0) {
      e.preventDefault();
      goBackToSearchField();
      setCurrentItemFocus(0);
    }
    if (e.key === 'Escape') goBackToSearchField();
  };

  const mapOptions = options.map((opt, i) => (
    <DropDownItem
      key={opt.uid}
      uid={opt.uid}
      label={opt.label}
      handleSelect={handleSelect}
      itemRef={(el) => {
        itemRefs.current[i] = el;
        // eslint-disable-next-line no-param-reassign
        if (i === 0) firstItemRef.current = el;
      }}
      handleKeyDown={(e) => handleItemKeyDown(e)}
      onFocus={() => setCurrentItemFocus(i)}
      isFocused={i === currentItemFocus}
    />
  ));
  // const mapOptions = <div>OPTIONS</div>

  const listDropdownClass = isOpen ? 'selectDropdown_list open' : 'selectDropdown_list';
  const menuDropdownClass = isOpen ? 'selectDropdown_menu open' : 'selectDropdown_menu';

  return (
    <div className="selectDropdown">
      <div className={menuDropdownClass} ref={menuRef}>
        {isOpen && <ul className={listDropdownClass}>{mapOptions}</ul>}
      </div>
    </div>
  );
};

CustomSelectDropdown.propTypes = {
  handleSelect: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
      label: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    })
  ).isRequired,
  isOpen: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired,
  firstItemRef: PropTypes.object.isRequired,
  goBackToSearchField: PropTypes.func.isRequired,
};

