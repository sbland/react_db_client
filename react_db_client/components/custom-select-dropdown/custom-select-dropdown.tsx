import React, { useEffect, useRef, useState } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';

import { DropDownItem } from './drop-down-item';
import { SelectDropdown, SelectDropdownList, SelectDropdownMenu } from './styles';

export interface Option {
  uid: string | number;
  label: string;
}

export interface ICustomSelectDropdownProps {
  options: Option[];
  handleSelect: (uid: string) => void;
  isOpen: boolean;
  handleClose: () => void;
  firstItemRef: React.MutableRefObject<HTMLElement | null>;
  goBackToSearchField: () => void;
  containerRef?: HTMLElement | null;
  position?: 'absolute' | 'relative';
  absolutePosition?: { left: number; top: number } | null;
}

export const CustomSelectDropdown: React.FC<ICustomSelectDropdownProps> = ({
  options,
  handleSelect,
  isOpen,
  handleClose,
  firstItemRef,
  goBackToSearchField = () => null,
  containerRef = null,
  position = 'relative',
  absolutePosition = null,
}: ICustomSelectDropdownProps) => {
  const menuRef = useRef<HTMLDivElement>(null);
  const itemRefs = useRef<HTMLElement[]>([]);
  const [currentItemFocus, setCurrentItemFocus] = useState(0);

  useEffect(() => {
    itemRefs.current = itemRefs.current.slice(0, options.length);
  }, [options]);

  function handleClickOutside(event) {
    if (menuRef.current && !menuRef.current.contains(event.target)) {
      handleClose();
    }
  }

  useEffect(() => {
    // Bind the event listener
    if (isOpen) {
      document.addEventListener('mousedown', handleClickOutside);
    } else {
      document.removeEventListener('mousedown', handleClickOutside);
    }
    return () => {
      // Unbind the event listener on clean up
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [isOpen]);

  const handleItemKeyDown = (e) => {
    if (e.key === 'ArrowDown' && currentItemFocus < options.length - 1) {
      e.preventDefault();
      itemRefs.current[currentItemFocus + 1].focus();
      setCurrentItemFocus((prev) => prev + 1);
    }
    if (e.key === 'ArrowUp' && currentItemFocus > 0) {
      e.preventDefault();
      itemRefs.current[currentItemFocus - 1].focus();
      setCurrentItemFocus((prev) => prev - 1);
    }
    if (e.key === 'ArrowUp' && currentItemFocus === 0) {
      e.preventDefault();
      goBackToSearchField();
      setCurrentItemFocus(0);
    }
    if (e.key === 'Escape') goBackToSearchField();
    if (e.key === 'Tab') {
      e.preventDefault();
      goBackToSearchField();
    }
  };

  const mapOptions = options.map((opt, i) => (
    <DropDownItem
      key={opt.uid}
      uid={opt.uid}
      label={opt.label}
      handleSelect={handleSelect}
      itemRef={(el) => {
        itemRefs.current[i] = el;
        // eslint-disable-next-line no-param-reassign
        if (i === 0) firstItemRef.current = el;
      }}
      handleKeyDown={(e) => handleItemKeyDown(e)}
      // onFocus={() => setCurrentItemFocus(i)}
      // isFocused={i === currentItemFocus}
    />
  ));
  // const mapOptions = <div>OPTIONS</div>

  const listDropdownClass = isOpen ? 'selectDropdown_list open' : 'selectDropdown_list';
  const menuDropdownClass = isOpen ? 'selectDropdown_menu open' : 'selectDropdown_menu';

  const render = (
    <SelectDropdown
      className="selectDropdown"
      style={{ position, left: absolutePosition?.left || 0, top: absolutePosition?.top || 0 }}
    >
      <SelectDropdownMenu className={menuDropdownClass} ref={menuRef}>
        {isOpen && (
          <SelectDropdownList className={listDropdownClass}>{mapOptions}</SelectDropdownList>
        )}
      </SelectDropdownMenu>
    </SelectDropdown>
  );
  if (containerRef) return ReactDOM.createPortal(render, containerRef);
  return render;
};

CustomSelectDropdown.propTypes = {
  handleSelect: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
      label: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    })
  ).isRequired,
  isOpen: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired,
  firstItemRef: PropTypes.object.isRequired,
  goBackToSearchField: PropTypes.func.isRequired,
  containerRef: PropTypes.any,
};

CustomSelectDropdown.defaultProps = {
  containerRef: null,
};
