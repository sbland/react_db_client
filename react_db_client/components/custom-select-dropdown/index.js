export { CustomSelectDropdown } from './custom-select-dropdown';
export { DropDownItem } from './drop-down-item';
