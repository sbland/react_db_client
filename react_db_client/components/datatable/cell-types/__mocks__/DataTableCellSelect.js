/* eslint-disable no-unused-vars */
/* eslint-disable arrow-body-style */
/* eslint-disable react/prop-types */
import React from 'react';

const DataTableCellSelect = ({
  cellData,
}) => {
  return (
    <div className="dataTableCellData dataTableCellData-select">
      {cellData}
    </div>
  );
};


export default DataTableCellSelect;
