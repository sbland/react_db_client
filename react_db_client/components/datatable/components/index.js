// export { DataTableCellHoverWrap } from './DataTableCell';
export { DataTableTotals } from './DataTableTotals';
export { DataTableHeadings } from './DataTableHeadings';
