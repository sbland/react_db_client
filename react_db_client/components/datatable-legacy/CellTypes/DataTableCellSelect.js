import React from 'react';
import PropTypes from 'prop-types';
import { CustomSelectDropdown } from '@react_db_client/components.custom-select-dropdown';
import { getRoot } from '@react_db_client/helpers.html-helpers';

/**
 * Data Cell Select
 *
 * @param {*} {
 *   cellData,
 *   updateData,
 * }
 * @returns
 */
const DataTableCellSelect = ({
  columnData: { options },
  cellData,
  acceptValue,
  resetValue,
  focused,
  editMode,
}) => {
  const acceptValueLocal = (v) => {
    acceptValue(v);
  };

  const rejectValue = () => {
    resetValue();
  };

  const displayValue = cellData && options && options.find((opt) => opt.uid === cellData)?.label;

  const containerRef = getRoot('selectContainer');
  return (
    <div className="dataTableCellData dataTableCellData-select">
      {/* TODO: Implement search dropdown */}
      {/* <div
        style={{
          display: (showEditor) ? 'block' : 'none',
        }}
        className="cellInput-select"
      >
        <SearchAndSelectDropdown
          searchFunction={dropDownSearch}
          handleSelect={acceptValueLocal}
          labelField="label"
          returnFieldOnSelect="uid"
          searchFieldTargetField="label"
        />
      </div> */}
      <div className="displayCellData dataTableCellData_text" style={{ position: 'absolute' }}>
        {displayValue}
      </div>
      <CustomSelectDropdown
        options={options}
        handleSelect={acceptValueLocal}
        isOpen={focused && editMode}
        firstItemRef={{ current: {} }}
        handleClose={() => rejectValue()}
        goBackToSearchField={() => rejectValue()}
        containerRef={containerRef}
      />
    </div>
  );
};

DataTableCellSelect.propTypes = {
  cellData: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  columnData: PropTypes.shape({
    options: PropTypes.arrayOf(
      PropTypes.shape({
        uid: PropTypes.string.isRequired,
        label: PropTypes.string.isRequired,
      })
    ).isRequired,
    readOnly: PropTypes.bool,
  }).isRequired,
  acceptValue: PropTypes.func.isRequired,
  resetValue: PropTypes.func.isRequired,

  focused: PropTypes.bool.isRequired,
  editMode: PropTypes.bool.isRequired,
};

DataTableCellSelect.defaultProps = {
  cellData: 0,
};

export default DataTableCellSelect;
