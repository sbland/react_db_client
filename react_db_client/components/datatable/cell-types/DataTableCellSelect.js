import React from 'react';
import PropTypes from 'prop-types';
import { CustomSelectDropdown } from '@react_db_client/components.custom-select-dropdown';
import { getRoot } from '@react_db_client/helpers.html-helpers';
import { DefaultCellInnerStyle } from './style';

/**
 * Data Cell Select
 *
 * @param {*} {
 *   cellData,
 *   updateData,
 * }
 * @returns
 */
const DataTableCellSelect = ({
  columnData: { options },
  cellData,
  acceptValue,
  resetValue,
  focused,
  editMode,
}) => {
  const targetRef = React.useRef(null);
  const [position, setPosition] = React.useState({ top: 0, left: 0 });

  const acceptValueLocal = (v) => {
    acceptValue(v);
  };

  const rejectValue = () => {
    resetValue();
  };
  const containerRef = getRoot('selectContainer');

  React.useEffect(() => {
    if (targetRef.current) {
      /* Setup the floating dropdown popup */
      const targetPos = targetRef.current.getBoundingClientRect();
      const containerPos = containerRef.getBoundingClientRect();

      const dropDownPos = {
        top: targetPos.y - containerPos.y + targetPos.height,
        left: targetPos.x - containerPos.x,
      };
      containerRef.style.cssText = `
        position: relative;
      `;

      setPosition(dropDownPos);
    }
  }, [containerRef]);

  const displayValue = cellData && options && options.find((opt) => opt.uid === cellData)?.label;

  return (
    <DefaultCellInnerStyle className="dataTableCellData dataTableCellData-select" ref={targetRef}>
      {/* TODO: Implement search dropdown */}
      {/* <div
        style={{
          display: (showEditor) ? 'block' : 'none',
        }}
        className="cellInput-select"
      >
        <SearchAndSelectDropdown
          searchFunction={dropDownSearch}
          handleSelect={acceptValueLocal}
          labelField="label"
          returnFieldOnSelect="uid"
          searchFieldTargetField="label"
        />
      </div> */}
      <div className="displayCellData dataTableCellData_text" style={{ position: 'absolute' }}>
        {displayValue}
      </div>
      <CustomSelectDropdown
        options={options}
        handleSelect={acceptValueLocal}
        isOpen={focused && editMode}
        firstItemRef={{ current: {} }}
        handleClose={() => rejectValue()}
        goBackToSearchField={() => rejectValue()}
        position="absolute"
        absolutePosition={position}
        containerRef={containerRef}
      />
    </DefaultCellInnerStyle>
  );
};

DataTableCellSelect.propTypes = {
  cellData: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  columnData: PropTypes.shape({
    options: PropTypes.arrayOf(
      PropTypes.shape({
        uid: PropTypes.string.isRequired,
        label: PropTypes.string.isRequired,
      })
    ).isRequired,
    readOnly: PropTypes.bool,
  }).isRequired,
  acceptValue: PropTypes.func.isRequired,
  resetValue: PropTypes.func.isRequired,

  focused: PropTypes.bool.isRequired,
  editMode: PropTypes.bool.isRequired,
};

DataTableCellSelect.defaultProps = {
  cellData: 0,
};

export default DataTableCellSelect;
