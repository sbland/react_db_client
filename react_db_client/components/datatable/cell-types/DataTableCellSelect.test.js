import React from 'react';
import { mount } from 'enzyme';
import { DataTableCellSelect } from './DataTableCellSelect';
import { CustomSelectDropdown } from '@react_db_client/components.custom-select-dropdown';

const updateData = jest.fn();
const acceptValue = jest.fn();
const resetValue = jest.fn();

const columnData = {
  type: 'select',
  options: [
    { uid: 'a', label: 'A' },
    { uid: 'b', label: 'B' },
    { uid: 'c', label: 'C' },
  ],
};

const defaultProps = {
  columnData,
  cellData: 'a',
  updateData,
  acceptValue,
  resetValue,
  rowData: {},
  editMode: false,
  focused: false,
};

// TODO: Reimplement select cell
describe('DataTableCellType - text', () => {
  let component;

  beforeEach(() => {
    updateData.mockClear();
    acceptValue.mockClear();
    resetValue.mockClear();
    component = mount(<DataTableCellSelect {...defaultProps} />);
  });

  const setEditMode = (c) => {
    c.setProps({ editMode: true, focused: true });
    c.update();
  };
  describe('Render', () => {
    test('should match snapshot', () => {
      const tree = component.debug();
      expect(tree).toMatchSnapshot();
    });
  });

  test('should show the cell data', () => {
    const displayCellData = component.find('.displayCellData');
    expect(displayCellData.text()).toEqual(columnData.options[0].label);
  });

  test('should open the selection dropdown when we enable editmode and is focused', () => {
    let selectDropdown = component.find(CustomSelectDropdown);
    expect(selectDropdown.props().isOpen).toEqual(false);
    setEditMode(component);
    selectDropdown = component.find(CustomSelectDropdown);
    expect(selectDropdown.props().isOpen).toEqual(true);
  });

  test.skip('should focus on first item in dropdown when in edit mode and is focused', () => {
    setEditMode(component);
    const selectDropdown = component.find(CustomSelectDropdown);
    expect(selectDropdown.props().isOpen).toEqual(true);
    const focusedElement = document.activeElement;
    expect(selectDropdown.getDOMNode()).toBe(focusedElement);
  });

  test.skip('should close selection dropdown on value accept', () => {
    setEditMode(component);
    let selectDropdown = component.find(CustomSelectDropdown);
    selectDropdown.props().handleSelect(columnData.options[0].uid);
    selectDropdown = component.find(CustomSelectDropdown);
    expect(selectDropdown.props().isOpen).toEqual(false);
  });

  // test('should unhide cell input when clicking the edit btn', () => {
  //   let cellInput = component.find(inputClassname);
  //   expect(cellInput).toBeTruthy();
  //   expect(cellInput.props().style.display).toEqual('none');
  //   const editBtn = component.find('.editBtn');
  //   editBtn.simulate('click');
  //   cellInput = component.find(inputClassname);
  //   expect(cellInput.props().style.display).toEqual('block');
  // });
  // test('should update data on cell data change', () => {
  //   setEditMode(component);
  //   const cellInput = component.find(inputClassname);
  //   const newValue = 'newentry';
  //   cellInput.simulate('change', { target: { value: newValue } });
  //   expect(updateData).toHaveBeenCalledWith(newValue);
  // });
  // test('should accept value on unfocus', () => {
  //   setEditMode(component);
  //   const cellInput = component.find(inputClassname);
  //   cellInput.simulate('blur');
  //   expect(acceptValue).toHaveBeenCalledWith(defaultProps.cellData);
  // });
  // test('should accept value on key down enter', () => {
  //   setEditMode(component);
  //   const cellInput = component.find(inputClassname);
  //   cellInput.simulate('keyDown', { key: 'Enter' });
  //   expect(acceptValue).toHaveBeenCalledWith(defaultProps.cellData);
  // });
  // test('should reset value to initial value if user presses escape', () => {
  //   setEditMode(component);
  //   const cellInput = component.find(inputClassname);
  //   const newValue = 'newentry';
  //   cellInput.simulate('change', { target: { value: newValue } });
  //   expect(updateData).toHaveBeenCalledWith(newValue);
  //   cellInput.simulate('keyDown', { key: 'Escape' });
  //   expect(resetValue).toHaveBeenCalledWith();
  // });
});
