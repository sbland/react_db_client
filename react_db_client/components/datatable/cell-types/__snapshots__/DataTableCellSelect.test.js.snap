// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTableCellType - text Render should match snapshot 1`] = `
"<DataTableCellSelect columnData={{...}} cellData=\\"a\\" updateData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} acceptValue={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} resetValue={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} rowData={{...}} editMode={false} focused={false}>
  <div className=\\"dataTableCellData dataTableCellData-select\\">
    <div className=\\"displayCellData dataTableCellData_text\\" style={{...}}>
      A
    </div>
    <CustomSelectDropdown options={{...}} handleSelect={[Function: acceptValueLocal]} isOpen={false} firstItemRef={{...}} handleClose={[Function: handleClose]} goBackToSearchField={[Function: goBackToSearchField]}>
      <div className=\\"selectDropdown\\">
        <div className=\\"selectDropdown_menu\\" />
      </div>
    </CustomSelectDropdown>
  </div>
</DataTableCellSelect>"
`;
